File: src/content/docs/topics/cs/coding-problems/linked-list/143-reorder-list-approach3.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def to_list(head):
    out = []
    while head:
        out.append(head.val)
        head = head.next
    return out

def from_list(vals):
    dummy = ListNode()
    cur = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next

def reorder_list(head) -> None:
    if not head or not head.next:
        return
    slow = fast = head
    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next
    prev, curr = None, slow.next
    slow.next = None
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    second = prev
    first = head
    while second:
        t1 = first.next
        t2 = second.next
        first.next = second
        second.next = t1
        first = t1
        second = t2

h = from_list([1,2,3,4]); reorder_list(h); assert to_list(h) == [1,4,2,3]
h = from_list([1,2,3,4,5]); reorder_list(h); assert to_list(h) == [1,5,2,4,3]
h = from_list([1]); reorder_list(h); assert to_list(h) == [1]
h = from_list([1,2]); reorder_list(h); assert to_list(h) == [1,2]
print("all tests pass")

File: src/content/docs/topics/cs/coding-problems/linked-list/143-reorder-list-approach3.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val: number = 0, next: ListNode | null = null) {
        this.val = val; this.next = next;
    }
}

function toList(head: ListNode | null): number[] {
    const out: number[] = [];
    while (head) { out.push(head.val); head = head.next; }
    return out;
}

function fromList(vals: number[]): ListNode | null {
    const dummy = new ListNode();
    let cur = dummy;
    for (const v of vals) { cur.next = new ListNode(v); cur = cur.next; }
    return dummy.next;
}

function reorderList(head: ListNode | null): void {
    if (!head || !head.next) return;
    let slow: ListNode = head;
    let fast: ListNode | null = head;
    while (fast.next && fast.next.next) {
        slow = slow.next!;
        fast = fast.next.next;
    }
    let prev: ListNode | null = null;
    let curr: ListNode | null = slow.next;
    slow.next = null;
    while (curr) {
        const nxt = curr.next;
        curr.next = prev;
        prev = curr;
        curr = nxt;
    }
    let first: ListNode | null = head;
    let second: ListNode | null = prev;
    while (second) {
        const t1 = first!.next;
        const t2 = second.next;
        first!.next = second;
        second.next = t1;
        first = t1;
        second = t2;
    }
}

let h = fromList([1,2,3,4]); reorderList(h); assert(JSON.stringify(toList(h)) === JSON.stringify([1,4,2,3]));
h = fromList([1,2,3,4,5]); reorderList(h); assert(JSON.stringify(toList(h)) === JSON.stringify([1,5,2,4,3]));
h = fromList([1]); reorderList(h); assert(JSON.stringify(toList(h)) === JSON.stringify([1]));
h = fromList([1,2]); reorderList(h); assert(JSON.stringify(toList(h)) === JSON.stringify([1,2]));
console.log('all tests pass');
